### PR TITLE
Adding Francisco Tapia's implementation of partial_sort

### PR DESCRIPTION
**Pull request by hkaiser**
@fjtapia thank you very much for contributing!

### CODE REVIEW

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/sort.hpp [hunk 1]
```diff
@@ -53,18 +54,20 @@ namespace hpx { namespace parallel { inline namespace v1 {
         /// \param iter_1 : iterator to the first value
         /// \param iter_2 : iterator to the second value
         /// \param iter_3 : iterator to the third value
-        /// \param comp : object for to compare two values
+        /// \param comp : object for to Comp two values
```
Comment (msimberg) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/sort.hpp:57]:
```suggestion
        /// \param comp : object for comparing two values
```

**Review by msimberg — APPROVED** [review]
Thanks, looks good to me!

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/partial_sort.hpp [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+                num = (num >> 4);
+            }
+            if (num >= (1ull << 2))
+            {
+                nb += 2;
+                num = (num >> 2);
+            }
+            if (num >= (1ull << 1))
+            {
+                nb += 12;
```
Comment (aurianer) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/partial_sort.hpp:132]: Shouldn't this be `1`?

**Review by aurianer — COMMENTED** [review]
Sorry always late :woman_facepalming:

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/partial_sort.hpp (hunk 2 above)
Comment (msimberg) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/partial_sort.hpp:132]: That's a very good catch! It sure looks like it.

Comment (hkaiser) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/partial_sort.hpp:132]: Will fix (straight to master). Thanks!

**Comment by fjtapia** [comment]
It is an error, must add 1.
Sorry

**Comment by fjtapia** [comment]
Hi Hartmut,
There is an error in the implementation of partial_sort_copy. The two first
iterators, the input range, are input iterator. With them only can
dereference and increment. The other two, the output range, are random
access.
If the input range would be random access iterators, the single and
parallel process can be done in place, without additional memory.
But in the specification talks about to remain unmodified the input range.
Then we have two scenarios
1.- When the input range is equal or small than the output range. We can
copy the input range in the output range and sort, in a single or parallel
mode.
2.- When the input range is greater than the output range, it needs an
auxiliary memory to copy the elements from the input range, and do a
partial_sort, single or parallel, and copy the results in the output range.

I am working on the modification of the code. In a few days, I will have
ready the new code.

About the fail in the function nbits, it is a consequence of a hurry.  I
use a meta programmed function, but is bigger, and wrote this function and,
against my custom. I didn't check and fail. My arrogance was defeated by
reality. I am sorry

Francisco

**Comment by hkaiser** [comment]
> It is an error, must add 1. Sorry

No problem, @aurianer caught it.

**Comment by hkaiser** [comment]
@fjtapia no worries, take your time. We have not added `partial_sort_copy` and `nth_element` to HPX yet. This PR adds `partial_sort` only.

**Comment by fjtapia** [comment]
THANKS

**Pull request merged**